hail/python/hail/vds/combiner/variant_dataset_combiner.py

```diff
@@ -0,0 +1,445 @@
+                json.dump(self, out, indent=2, cls=Encoder)
+            fs.remove(backup_path)
+        except OSError as e:
+            # these messages get printed, because there is absolutely no guarentee
+            # that the hail context is in a sane state if any of the above operations
+            # fail
+            print(f'Failed saving {self.__class__.__name__} state at {self.save_path}')
+            print(f'An attempt was made to copy {self.save_path} to {backup_path}')
+            print('An old version of this state may be there.')
+            print('Dumping current state as json to standard output, you may wish '
+                  'to save this output in order to resume the combiner.')
+            json.dump(self, sys.stdout, indent=2, cls=Encoder)
+            raise e
+
+    @staticmethod
+    def load(path) -> 'VariantDatasetCombiner':
+        fs = hl.current_backend().fs
+        with fs.open(path) as stream:
+            combiner = json.load(stream, cls=Decoder)
+            if combiner.save_path != path:
+                warning('path/save_path mismatch in loaded VariantDatasetCombiner, using '
+                        f'{path} as the new save_path for this combiner')
+                combiner.save_path = path
+            return combiner
+
+    def to_dict(self) -> dict:
+        intervals_typ = hl.tarray(hl.tinterval(hl.tlocus(self.reference_genome)))
+        return {'name': self.__class__.__name__,
+                'save_path': self.save_path,
+                'output_path': self.output_path,
+                'temp_path': self.temp_path,
+                'reference_genome': str(self.reference_genome),
+                'branch_factor': self.branch_factor,
+                'target_records': self.target_records,
+                'gvcf_batch_size': self.gvcf_batch_size,
+                'gvcf_external_header': self.gvcf_external_header,  # put this here for humans
+                'contig_recoding': self.contig_recoding,
+                'vdses': self.vdses,
+                'gvcfs': self.gvcfs,
+                'gvcf_sample_names': self.gvcf_sample_names,
+                'gvcf_import_intervals': intervals_typ._convert_to_json(self.gvcf_import_intervals),
+                }
+
+    def step(self):
+        if self.finished:
+            return
+        if self.gvcfs:
+            self._step_gvcfs()
+        else:
+            self._step_vdses()
+        if not self.finished:
+            self._job_id += 1
+
+    def _step_vdses(self):
+        files_to_merge = self.vdses[-self.branch_factor:]
+        self.vdses = self.vdses[:-self.branch_factor]
+
+        temp_path = self._temp_out_path(f'vds-combine_job{self._job_id}')
+        largest_vds = files_to_merge[0]
+        vds = hl.vds.read_vds(largest_vds.path)
+        # we use the reference data since it generally has more rows than the variant data
+        intervals, intervals_dtype = calculate_new_intervals(vds.reference_data,
```

Comment:
This is going to get executed for each VDS combine 'job', rather than once per 'phase'. How can we avoid this? Maybe we can cache the intervals in buckets that refer to ranges of sample sizes?

Reply:
see 89f66e2

Reply:
beautiful.